datafusion/common/src/tree_node.rs

```diff
@@ -18,7 +18,6 @@
 //! This module provides common traits for visiting or rewriting tree
 //! data structures easily.
 
-use std::borrow::Cow;
```

Comment:
It seems like this PR reverts @viirya's https://github.com/apache/arrow-datafusion/pull/8672. Why is that?

Reply:
@viirya's PR was a good step towards deduplicating the `apply_children()` implementations. This PR further deduplicates not only `apply_children()`, but also `map_children()`, bringing them together in one place. 

Like the `children()` method introduced by @viirya, which returns a `Cow`, `map_children()` requires two different methods: one to separate the children from self for mutation, and another for reattaching them afterwards. Since obtaining ownership of the children without owning self is not possible, I think `Cow`'s purpose becomes not much relevant in this use case. Now we have two children related methods: one with a reference and one with ownership. The original `fn children(&self) -> Vec<Cow<Self>>` can still be used, but it has become redundant. If the user requires write access, `take_children()` should be used.

Rather than implementing functions such as `apply_children()` and `map_children()`, which provide too much flexibility and are prone to design an unclear flow, I think it is more appropriate to implement these 3 methods, which express the relationships of self nodes with their children, both in terms of ease of use and simplicity.

Reply:
If you check `apply_children()`, it seems like you duplicates its logic to 4 places in this PR.
Also the `Cow` was there to avoid cloning in `Expr` but now we have cloning again.
I'm ok with the new `children(&self)` and `take_children(self)` methods, but it seems like they are not properly used here and there...

Reply:
> If you check `apply_children()`, it seems like you duplicate its logic to 4 places in this PR.\\

Right. The point is that there is a fixed number of `apply_children` implementations (one doesn't need to implement it per rule or per use case). However, 1 is obviously better than 4 and we will iterate on this.

> Also, the `Cow` was there to avoid cloning in `Expr` but now we have cloning again.

Yes, and I am looking at ways to eliminate it. This is the only use case remaining for`Cow`, and there may be a good way to remove it and still avoid cloning. The strategy in the WIP stage is to first explicitly implement everything and then progressively deduplicate by refining design. Maybe we will converge to the same `Cow` design, in that case we will simply revert.

Thanks for taking a look 🚀